Monad add prop from fix

zeffii opened the pull request:
addresses a few longstanding incomplete features in monad.

zeffii commented:
almost!! 

i know what to do now.   but first some sunshine.

zeffii commented:
almost almost.. i think the reason the socket.prop_name is not set for snlite node is because the socket acquisition code in nodes/scene/monad.py  is not tuned for nodes with 'prop_type'.

Durman commented on nodes/scene/monad.py:
```diff
@@ -84,15 +84,20 @@ def update(self):
+                    new_socket.prop_name = prop_name
+                else:
+                    for name, value in prop_data.items():
+                        if not name == 'prop_name':
```
```suggestion
                        if name != 'prop_name':
```

zeffii commented:
many issues remain, i'll add an 'ensure unique prop name` function, and i think the result is already an improvement. many issues remain, but commenting this complicated code was useful.

zeffii commented:
something like this
```
import re

node = lambda: None
node.__annotations__ = {}

node.__annotations__["floats_1_amp"] = lambda: None
node.__annotations__["floats_2_degree"] = lambda: None
node.__annotations__["floats_3_amp2"] = lambda: None
node.__annotations__["floats_4_float"] = lambda: None
node.__annotations__["floats_5_rads"] = lambda: None
node.__annotations__["ints_1_numrings"] = lambda: None
node.__annotations__["ints_2_numrings"] = lambda: None
node.__annotations__["ints_3_fixer"] = lambda: None
node.__annotations__["ints_4_frogs"] = lambda: None
node.__annotations__["ints_5_count"] = lambda: None

def get_number(new_prop_name, regex):
    matches= re.search(regex, new_prop_name)
    return int(matches.group('number'))

def unique_name_test(obj_id, new_prop_name):

    # can be used unchanged
    if not (new_prop_name in obj_id.__annotations__):
        return new_prop_name

    regex = r"^(?P<kind>ints|floats)_(?P<number>\d+)_(?P<variable>\S+)$"
    
    matches= re.search(regex, new_prop_name)
    kind = matches.group('kind')
    num = matches.group('number')
    variable = matches.group('variable')
    # print(f"kind {kind} : num {num}")
    
    # find the highest token and return it plus one
    names = list(filter(lambda n: n.startswith(f"{kind}_"), obj_id.__annotations__))
    new_num = max([get_number(n, regex) for n in names]) + 1
    proposed_new_name = f"{kind}_{new_num}_{variable}"
    return proposed_new_name

unique_name_test(node, 'ints_3_fixer')
# >>> ints_6_fixer
```

zeffii commented:
maybe..
```python
import re

node = lambda: None
node.__annotations__ = {}

node.__annotations__["floats_1_amp"] = lambda: None
node.__annotations__["floats_2_degree"] = lambda: None
node.__annotations__["floats_3_amp2"] = lambda: None
node.__annotations__["floats_4_float"] = lambda: None
node.__annotations__["floats_5_rads"] = lambda: None
node.__annotations__["ints_1_numrings"] = lambda: None
node.__annotations__["ints_2_numrings"] = lambda: None
node.__annotations__["ints_3_fixer_more"] = lambda: None
node.__annotations__["ints_4_frogs"] = lambda: None
node.__annotations__["ints_5_count"] = lambda: None

def get_number(new_prop_name):
    regex = r"^(?P<kind>ints|floats)_(?P<number>\d+)_(?P<variable>\S+)$"
    matches= re.search(regex, new_prop_name)
    return int(matches.group('number'))

def unique_name_test(obj_id, new_prop_name):

    # can be used unchanged
    if not (new_prop_name in obj_id.__annotations__):
        return new_prop_name
   
    # find the highest token and return it plus one
    kind, num, variable = new_prop_name.split('_', 2)

    names = list(filter(lambda n: n.startswith(f"{kind}_"), obj_id.__annotations__))
    new_num = max([get_number(n) for n in names]) + 1
    return f"{kind}_{new_num}_{variable}"


print(unique_name_test(node, 'ints_3_fixer_more'))
```

zeffii commented:
...maybe
```python
node = lambda: None
node.__annotations__ = {}

node.__annotations__["floats_1_amp"] = lambda: None
node.__annotations__["floats_2_degree"] = lambda: None
node.__annotations__["floats_3_amp2"] = lambda: None
node.__annotations__["floats_4_float"] = lambda: None
node.__annotations__["floats_5_rads"] = lambda: None
node.__annotations__["ints_1_numrings"] = lambda: None
node.__annotations__["ints_2_numrings"] = lambda: None
node.__annotations__["ints_3_fixer_more"] = lambda: None
node.__annotations__["ints_4_frogs"] = lambda: None
node.__annotations__["ints_5_count"] = lambda: None


def unique_name_test(obj_id, new_prop_name):

    # can be used unchanged
    if not (new_prop_name in obj_id.__annotations__):
        return new_prop_name
   
    kind, num, variable = new_prop_name.split('_', 2)

    # find the highest token and return it plus one
    name = sorted(list(filter(lambda n: n.startswith(f"{kind}_"), obj_id.__annotations__)))[-1]
    _, num, _ = name.split('_', 2)
    return f"{kind}_{int(num) + 1}_{variable}"


print(unique_name_test(node, 'ints_3_fixer_more'))
```

zeffii commented:
ok.. settling on this because it makes sure the list of named props match the signature `<kind>_<num><varname>`

```python
import re

node = lambda: None
node.__annotations__ = {}

node.__annotations__["floats_1_amp"] = lambda: None
node.__annotations__["floats_2_degree"] = lambda: None
node.__annotations__["floats_3_amp2"] = lambda: None
node.__annotations__["floats_4_float"] = lambda: None
node.__annotations__["floats_5_rads"] = lambda: None
node.__annotations__["ints_1_numrings"] = lambda: None
node.__annotations__["ints_2_numrings"] = lambda: None
node.__annotations__["ints_3_fixer_more"] = lambda: None
node.__annotations__["ints_4_frogs"] = lambda: None
node.__annotations__["ints_5_count"] = lambda: None
node.__annotations__["ints_upsula"] = lambda: None
node.__annotations__["floats_steiner"] = lambda: None
node.__annotations__["floatslister"] = lambda: None
node.__annotations__["float_4_4_lister"] = lambda: None


def find_all_matching_params(obj_id, kind):
    regex = "^(?P<kind>ints|floats)_(?P<number>\d+)_(?P<variable>\S+)$"
    names = []
    for name in obj_id.__annotations__:
        matches = re.search(regex, name)
        if matches and matches.group('kind') == kind:
            names.append(name)
    return names


def unique_name_test(obj_id, new_prop_name):

    # can be used unchanged
    if not (new_prop_name in obj_id.__annotations__):
        return new_prop_name
   
    kind, num, variable = new_prop_name.split('_', 2)

    # find the highest token and return it plus one
    names = find_all_matching_params(obj_id, kind)
    name = sorted(names)[-1]
    _, num, _ = name.split('_', 2)
    return f"{kind}_{int(num) + 1}_{variable}"


print(unique_name_test(node, 'ints_3_fixer_more'))

```

Durman commented on core/monad_properties.py:
```diff
@@ -186,6 +187,32 @@ def get_settings(self):
+
+def ensure_unique(obj_id, new_prop_name):
+
+    # can be used unchanged
+    if not (new_prop_name in obj_id.__annotations__):
+        return new_prop_name
```
Why don't use `not in` operator?

zeffii replied:
you mean 
```python
if new_prop_name not in annotations:
        return new_prop_name
```
i'll think about it. but it's a minor issue

zeffii commented:
i've spent time with this code again, and i think i'm "friends" with the monad codebase.  sadly there are still a few pesky issues
- updates being triggered in nodetree before socket gets data
- adding two A number nodes in Float mode, and then removin the first socket (using the X ) will result in both sockets accessing the same `prop_name` (my effort here to devise the `ensure_unique_name` function is clearly misplaced)

not going to psh this yet.. but more progress might come tomorrow.

zeffii commented:
those main issues are now resolved. My main remaining issue is a few loose updates that trigger the node before socket data is ready.

zeffii commented:
this makes the stdout a little noisy.
